### PR TITLE
fsrepo migrations

### PR DESCRIPTION
**Pull request by jbenet**
This PR adds migrations for fsrepo.
- a very simple migration structure
- a 0-to-1 migration adding the version file
- a 1-to-2 migration moving things around to look more
  like https://github.com/ipfs/specs/tree/master/repo/fs-repo

The 1-to-2 isn't done yet and will be the meat of this PR. Need
- [x] flat repo in `blocks/`
- [x] move repo lock (not needed in the migration, just code)
- [ ] make api file (not needed in the migration, just code)
- [ ] make hooks dir
- [x] incorporate .ipfs change (#975)

### CODE REVIEW

**Comment by jbenet** [comment]
@whyrusleeping not sure how you want to store things in keyspec, but is the layout in https://github.com/ipfs/specs/tree/master/repo/fs-repo#keys work for you? note the `.{pub, pri, sym}` endings.

**Comment by whyrusleeping** [comment]
@jbenet thats a little non-standard (i was going for ssh-compliant naming and formats) but I can work with it.

**Comment by jbenet** [comment]
see discussion in https://github.com/ipfs/go-ipfs/pull/975#issuecomment-87483806

**Comment by whyrusleeping** [comment]
@jbenet why do you have a binary commited?

**Comment by jbenet** [comment]
@whyrusleeping because i fail. thanks

- [ ] remove `0-to-1` binary

**Comment by whyrusleeping** [comment]
lets make sure it gets GCed out! 2MB makes a difference to new clones

**Comment by whyrusleeping** [comment]
this looks good to me, i like the approach.

**Comment by whyrusleeping** [comment]
@jbenet should the daemon check the repo version on startup, and run the appropriate migration tools? (or, prompt the user to run them)

**Comment by jbenet** [comment]
@whyrusleeping for now prompt the user to run them. we can do things automatically later, but if something goes wrong for now, would be ideal to have the user on top of it

**Comment by whyrusleeping** [comment]
im not convinced we should have a tool for each migration, i think a single tool would be the better approach

**Comment by jbenet** [comment]
Need to make sure we can rest assured these will work years from now without our maintenance. The repo's code will progress, often change fundamentally. Forcing us to maintain things backwards compatible with all repo versions -- or even one tool -- is signing us up for pain in the future, testing with old things, etc. Instead, if we freeze its code now, with its own vendoring, we can count on these working with the current repo versions down the road. 

I'm envisioning a pretty simple setup where:

```
fs-repo-migrations/
├── 0-to-1/
│   ├── all_the_code_needed_for_0-to-1/
│   └── main.go
├── 1-to-2/
│   ├── all_the_code_needed_for_1-to-2/
│   └── main.go
└── 2-to-3/
    ├── all_the_code_needed_for_2-to-3/
    └── main.go
```
so we can go into any of these, compile and know it will work. 

depending on how we do things, we may not need to use godeps, we could just vendor directly (cp -r) into those directories.

File: repo/fsrepo/fsrepo.go [hunk 1]
```diff
@@ -102,6 +112,24 @@ func open(repoPath string) (repo.Repo, error) {
...
+		if os.IsNotExist(err) {
+			return nil, errors.New("version check failed, no version file found, please run 0-to-1 migration tool.")
+		}
+		return nil, err
+	}
+
+	vers := string(ver)[:1]
+
+	if vers != RepoVersion {
+		return nil, fmt.Errorf("Repo has incorrect version: '%s'\nProgram version is: '%s'\nPlease run the appropriate migration tool before continuing",
```
Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: var MsgUpgradeRepo = `Repo has incorrect version: %s
Program version is: %s
Please run the repo migration tool before continuing:

    <cmd for them to run>
`

I'll think about what a good way to do `<cmd for them to run>` is. the problem is they:
- may not have it installed (installed ipfs as a binary).
- may not have go installed (may not be able to compile).
- do not have ipfs installed (cannot `ipfs cat <hash>/<arch>/1-to-2 >1-to-2`)


May be able to do something like:
```sh
wget http://gateway.ipfs.io/ipfs/<hash>/<arch>/1-to-2
chmod +x 1-to-2
./1-to-2
```

And ship a little shell script that does it as:

```sh
ipfs-repo-migrate 1-to-2
```

Comment (jbenet) [inline repo/fsrepo/fsrepo.go]: go 1.5 is coming at such a good moment...

Comment (whyrusleeping) [inline repo/fsrepo/fsrepo.go]: we could at least just have a page somewhere discussing migrations, maybe a github repo, and link them to it in the error message

**Comment by jbenet** [comment]
@whyrusleeping LGTM!! I'll write the script to download the migrations for us.

File: cmd/ipfs/daemon.go [hunk 2]
```diff
@@ -104,23 +104,18 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
 		}
 	}
 
-	// To ensure that IPFS has been initialized, fetch the config. Do this
-	// _before_ acquiring the daemon lock so the user gets an appropriate error
-	// message.
-	// NB: It's safe to read the config without the daemon lock, but not safe
-	// to write.
```
Comment (jbenet) [inline cmd/ipfs/daemon.go]: @whyrusleeping will this still yield the right error?

**Comment by jbenet** [comment]
@whyrusleeping the new changes may alter the semantics of ipfs startup. are the right errors still gotten?

I'll start full tests manually.

File: cmd/ipfs/daemon.go (hunk 2 above)
Comment (whyrusleeping) [inline cmd/ipfs/daemon.go]: the errors returned may now differ, but they actually mean something now. the other error was completely worthless

**Comment by jbenet** [comment]
Every commit passed for me. And i think for @whyrusleeping as well.

**Comment by jbenet** [comment]
@whyrusleeping what's remaining here? everything so far LGTM.

---

I can see we need a way for people to download + run the migration. This is going to be a bit annoying, given that it's a different binary and people may have installed from prebuilt bins.

I think we can make a better system down the road, but for now maybe:

- add prebuilt binaries of each tool for each arch
- make a script that downloads the right binary for people given their arch. 

thoughts?

**Comment by whyrusleeping** [comment]
I think we can probably get away with using gobuilder (and maybe rehosting into ipfs). I can probably update my error messages to point to a 'migrations readme' page online.

**Comment by jbenet** [comment]
@whyrusleeping i'm making a page now. I think we should do these migrations in a separate repo, here's why:

- code isolation from go-ipfs, can be properly frozen
- can be easily used by other implementations
- easier to navigate / look at for people looking at what migrations will do
- it will give us a bit of experience working with multiple go repos under the org

thoughts?

**Comment by whyrusleeping** [comment]
Agreed. In writing `1-to-2` i tried to avoid using go-ipfs code when i could (minus a few repo imports) to make this easier.

**Comment by jbenet** [comment]
> i tried to avoid using go-ipfs code when i could (minus a few repo imports) to make this easier.

+1. We should probably vendor those just in case :| -- we dont have to use godeps. we can just copy-paste. (:open_mouth:!)

**Comment by whyrusleeping** [comment]
im fine with copy pasting, especially since we are 'vendoring' with the explicit goal of never changing that code

**Comment by jbenet** [comment]
@whyrusleeping what's missing here? just the guide to upgrade?

**Comment by whyrusleeping** [comment]
Just the guide.

**Comment by jbenet** [comment]
### Migrations CR

#### - [ ] error needs to be friendlier

now:
```
Error: ipfs repo found in old '.go-ipfs' location, please run migration tool
```

should be something like:

```
Error: ipfs repo found in old '.go-ipfs' location, please run migration tool:

    wget http://gateway.ipfs.io/ipfs/<hash>/<go-platform>/ipfs-fsrepo-migrations
    chmod +x ipfs-fsrepo-migrations
    ./ipfs-fsrepo-migrations

If you encounter problems, please see https://github.com/ipfs/fs-repo-migrations/blob/master/run.md
```
This should work _on every machine_ with as few steps as possible. We _could_ detect if Go is installed, and suggest using `go get`, but this should only be if they do have it installed... we should not have any `go get` crap if the machine doesn't have go. We _could_ also use a shell script and pipe straight into shell, but they dont have hashpipe installed and i dont want to do that if they don't.

It's really funny, if ipfs worked during this process, or could install things from it, this would be so easy. lol :chicken: :egg: 

#### - [ ] running the migration notes important info

see https://github.com/ipfs/fs-repo-migrations/issues/1

#### - [ ] make https://github.com/ipfs/fs-repo-migrations/blob/master/run.md good.

i'm on this

#### - [x] migration worked!! \o/
#### - [ ] migration fails mysteriously if daemon running

https://github.com/ipfs/fs-repo-migrations/issues/2

**Comment by jbenet** [comment]
On the message, maybe we can say:

```
Error: ipfs repo found in old '.go-ipfs' location, please run migration tool.
Please see https://github.com/ipfs/fs-repo-migrations/blob/master/run.md
Sorry for the inconvenience. In the future, these will run automatically. :)
```

**Comment by jbenet** [comment]
- [ ] mv `daemon.lock -> repo.lock`

**Comment by jbenet** [comment]
rebased on new master

**Pull request merged**